homeassistant/components/xiaomi_miio/config_flow.py

```diff
@@ -149,4 +161,29 @@ async def async_step_device(self, user_input=None):
         else:
             schema = DEVICE_CONFIG
 
+        if errors:
+            schema.extend(DEVICE_MODEL_CONFIG)
+
         return self.async_show_form(step_id="device", data_schema=schema, errors=errors)
+
+    async def async_get_mac(self, device_info, host):
+        """Get the mac address of a Miio Device when the info call fails."""
+        if device_info is not None:
+            return format_mac(device_info.mac_address)
+
+        # If the info call failed, use getmac as backup
+        try:
+            mac = await self.hass.async_add_executor_job(
+                partial(get_mac_address, **{"ip": host})
```

Comment:
Getting the mac address from the ip address won't work on all networks. We try to avoid using this.

Can we instead use a form and ask the user for the required connection details?

Reply:
The get_mac_address is normally not used.
Only if the `info()` call to the device fails (to get the mac and model) it will be used as a backup.

All Miio Devices will normally respond to the info() call.
Only if a users blocks the Miio device from accesing the internet (only local traffic allowed) on some minority of the devices (only 1 type of a vacuum known to have this issue) the info() call will fail.
Almost all Miio Devices will just respond to the info() call even if you block it's internet acces.

So bassically the get_mac will only be used in some edge cases as a last last last resort.
(Only known situation will be a gen 1 rockrobo vacuum and blocking access to the Internet)

To avoid the get_mac call, we would need to ask the user for the mac_address. That will then be directly used as the unique_id of the device.
We can do that but if a user supplies a wrong mac_address the device could be discovered multiple times (and set up multiple times), this is such an edge case that I don't really mind tho...

Reply:
@MartinHelmare what do you think?
Use get_mac or use a form to ask the user for the mac_address?

Reply:
It looks like we only use the mac address for unique id. Then I suggest we don't set a unique id if we're missing mac address.

Reply:
It is also used for all the entity unique id's.
How does that work if you don't set a unique_id?
can you just supply the value `None`?

What effects will that have?

Reply:
Yes, `None` is the default so that works. It will mean that none of the entity registry features will be active for those entities, like changing name in the GUI per entity.

Reply:
@MartinHjelmare thanks!

I just made a commit to let the config flow proceed even if the mac cannot be retrieved (with value None).
So now the order of the source from which the mac is used:

1) zeroconf discovery info
2) from the info() call to the device
3) getmac - ip
4) getmac - hostname
5) pass None

I kept the getmac if it is okay with you, but it has a broad try: except: around it and if it fails the config flow will still continue with passing None, so it is just an extra "backup" attempt to still retrieve the mac.

In the vast majority of situation option 1) or 2) will be used to get the mac.

Reply:
I think we should remove the use of `getmac` completely.

Reply:
I removed getmac.